### PR TITLE
[ML] Better handling of unbalanced class representation in the training data

### PR DESCRIPTION
**Pull request by tveasey**
This adds two features:
1. Stratifying the sampling we do for each test/train fold for classification.
2. Implementing a reweighting scheme so that each class contributes approximately equal weight to the training loss.

We sample uniformly at random from the training rows for each class in proportion to their frequency in the full train set. We then optionally reweight the loss for each example by the inverse frequency of its category in the training set.

### CODE REVIEW

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 1]
```diff
@@ -246,7 +246,7 @@ void CBoostedTreeTest::testPiecewiseConstant() {
             0.0, modelBias[i][0],
             4.0 * std::sqrt(noiseVariance / static_cast<double>(trainRows)));
         // Good R^2...
-        CPPUNIT_ASSERT(modelRSquared[i][0] > 0.96);
+        CPPUNIT_ASSERT(modelRSquared[i][0] > 0.95);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: Good question! This is because I pass the random number generator by value to the function which generates the cross validation row masks. Before, we were using it directly and so we get different random numbers generated after this point. This happens very early during training, so affects most of random numbers generated.
